pkg/loadbalancer/experimental/services.go

```diff
@@ -0,0 +1,394 @@
+// Copyright Authors of Cilium
+
+package experimental
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/reconciler"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/cilium/cilium/pkg/container"
+	"github.com/cilium/cilium/pkg/loadbalancer"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+// Services provides validated write access to the frontend and backend tables.
+//
+// This is currently experimental and not in use by default.
+// See comment in [tables.go] for further context.
+type Services struct {
+	log *slog.Logger
+	db  *statedb.DB
+	fes statedb.RWTable[*Frontend]
+	bes statedb.RWTable[*Backend]
+
+	hooks []ServiceHook
+}
+
+func NewServices(cfg Config, log *slog.Logger, db *statedb.DB, fes statedb.RWTable[*Frontend], bes statedb.RWTable[*Backend]) (*Services, error) {
+	if !cfg.EnableNewServices {
+		return nil, nil
+	}
+	return &Services{
+		log:   log,
+		db:    db,
+		fes:   fes,
+		bes:   bes,
+		hooks: []ServiceHook{},
+	}, nil
+}
+
+func (s *Services) IsEnabled() bool {
+	return s != nil
+}
+
+// Hook is a function that runs when a service is upserted. It can perform reads using the provided
+// ReadTxn, but for consistency it must not depend on any other data (except when the data is constant).
+// The hook can modify the service object.
+type ServiceHook func(statedb.ReadTxn, *Frontend)
+
+type ServiceWriteTxn struct {
+	statedb.WriteTxn
+}
+
+var (
+	ErrFrontendSourceMismatch = errors.New("frontend exists with different source")
+	ErrFrontendConflict       = errors.New("conflict with a service sharing address with a different name")
+)
+
+// RegisterHook adds a new hook to modify a service when upserted. This must only be called from
+// providers/invokes and not after starting!
+func (s *Services) RegisterHook(h ServiceHook) {
+	s.hooks = append(s.hooks, h)
+}
+
+// Frontends returns the frontend table for reading.
+// Convenience method for reducing dependencies.
+func (s *Services) Frontends() statedb.Table[*Frontend] {
+	return s.fes
+}
+
+// Services returns the backend table for reading.
+// Convenience method for reducing dependencies.
+func (s *Services) Backends() statedb.Table[*Backend] {
+	return s.bes
+}
+
+// WriteTxn returns a write transaction against services & backends and other additional
+// tables to be used with the methods of [Services]. The returned transaction MUST be
+// Abort()'ed or Commit()'ed.
+func (s *Services) WriteTxn(extraTables ...statedb.TableMeta) ServiceWriteTxn {
+	return ServiceWriteTxn{
+		s.db.WriteTxn(s.fes, append(extraTables, s.bes)...),
+	}
+}
+
+func (s *Services) UpsertFrontend(txn ServiceWriteTxn, params *FrontendParams) (created bool, err error) {
+	var fe Frontend
+
+	existing, _, found := s.fes.Get(txn, FrontendL3n4AddrIndex.Query(params.Address))
+	if found {
+		// Do not allow frontends with the same address but different names to override each other.
```

Comment:
I'm a bit unsure about these validations. Should we consider just dropping them at this point?

Reply:
Yeah, good idea!